src/idpf.rs

```diff
@@ -127,34 +125,51 @@ where
     }
 }
 
+/// Trait for values to be programmed into an IDPF.
+///
+/// Values must form an Abelian group, so that they can be secret-shared, and the group operation
+/// must be represented by [`Add`]. An implementation of [`CoinToss`] must be provided to randomly
+/// select a value using PRG output. Values must be encodable and decodable, without need for a
+/// decoding parameter.
+pub trait IdpfValue:
+    Add<Output = Self>
+    + AddAssign
+    + Sub<Output = Self>
+    + ConditionallySelectable
+    + ConditionallyNegatable
+    + CoinToss
+    + Encode
+    + Decode
+    + Sized
+{
+    /// Returns the additive identity.
+    fn zero() -> Self;
+}
+
 /// An output from evaluation of an IDPF at some level and index.
 #[derive(Debug, PartialEq, Eq)]
-pub enum IdpfOutputShare<const OUT_LEN: usize, FI, FL> {
+pub enum IdpfOutputShare<VI, VL> {
     /// An IDPF output share corresponding to an inner tree node.
-    Inner([FI; OUT_LEN]),
+    Inner(VI),
     /// An IDPF output share corresponding to a leaf tree node.
-    Leaf([FL; OUT_LEN]),
+    Leaf(VL),
 }
 
-impl<const OUT_LEN: usize, FI, FL> IdpfOutputShare<OUT_LEN, FI, FL>
+impl<VI, VL> IdpfOutputShare<VI, VL>
 where
-    FI: FieldElement,
-    FL: FieldElement,
+    VI: IdpfValue,
+    VL: IdpfValue,
 {
-    /// Combine two output share vectors into one.
-    pub fn merge(self, other: &Self) -> Result<IdpfOutputShare<OUT_LEN, FI, FL>, IdpfError> {
+    /// Combine two output share values into one.
+    pub fn merge(self, other: &Self) -> Result<IdpfOutputShare<VI, VL>, IdpfError> {
         match (self, other) {
-            (IdpfOutputShare::Inner(mut self_array), IdpfOutputShare::Inner(other_array)) => {
-                for (dest, src) in self_array.iter_mut().zip(other_array.iter()) {
-                    *dest += *src;
-                }
-                Ok(IdpfOutputShare::Inner(self_array))
+            (IdpfOutputShare::Inner(mut self_value), IdpfOutputShare::Inner(other_value)) => {
+                self_value += *other_value;
```

Comment:
If the method takes `Self`, you still have an implicit copy because you have to move `other` into the function. Or are you saying you'd prefer the copy be explicit by putting it in the function signature?

Reply:
I think I may be able to eliminate the copy by requiring `AddAssign<Rhs=&Self>` instead of `AddAssign<Rhs=Self>`, and then using that here. I'll tinker with it.

Reply:
No, in my head I though it *could* be more efficient to have merge() consume `other_value`. Might this be true of it's a big vector? LIke imagine an IDPF whose values are 32 elements long. We probably won't ever have this, but it's theoretically possible to use it this way.

Reply:
FWIW this is the first time we have `Add` for an array of field elements in this library.